docs/apis/subsystems/output/index.md

```diff
@@ -198,66 +198,76 @@ For each of the functions below we'll try to explain when they should be used, e
 
 ### String formatting functions
 
+The format_string and format_text functions should always be used when preparing the output of information. They may also be used to process information before it is stored in the database however, filters should only be applied at output. For example, language filters must only be applied as the content is prepared for output because we don't yet know the user's preferred language.
+
 #### p() and s()
 
 ```php
 function s($var, $strip=false)
 function p($var, $strip=false)
 ```
 
-These functions share the same code, so they will be explained together. The only difference is that `s()` returns the string, while `p()` prints it directly.
+The only difference between these two functions is that `s()` returns the string, while `p()` prints it directly.
 
 These functions should be used to:
 
 - print all the **values of form fields** like `<input>` or `<textarea>` tags.
 - **show plain (non-HTML) text** that has been introduced by the user (search string, quiz responses, ...).
-- print, in general, all the **dynamic data, not being HTML**, that doesn't need to be cleaned nor processed by filters. It is important not to use these functions for strings that contain HTML markup.
+- print, in general, all **dynamic data that is not HTML**, does not need to be cleaned or processed by filters. Never use these functions for strings that contain HTML markup.
 
-The functions replace certain characters that would have special meaning in HTML (`<, >, ", ', and &`) by HTML entities, so that they are displayed as intended. Note that even though the value of form fields printed with `p()` will have these characters converted to HTML entities, the submitted values will still contain the original characters.
+The functions replace certain characters that have a special meaning in HTML (`<, >, ", ', and &`) with HTML entities so that they are displayed as intended. Note that even though the value of form fields printed with `p()` will have these characters converted to HTML entities, the submitted values will still contain the original characters.
 
 The key parameter for this function is:
 
-- **strip**: it decides if we want to strip slashes from the string or not. By default, it's `false`, so no strip will be performed. We should set such parameter to 'true' only when data to be processed isn't coming from database, but from HTTP requests (forms, links, ...).
+- **strip**: Set to `true` to strip slashes from the string. Only set this parameter to `true` when the data to be processed is not coming from the database. This should be used when the string comes from HTTP requests (forms, links, ...). (Default is `false`, so no strip will be performed)
 
 #### format_text()
 
+```php
+function format_text(string $text, $format = FORMAT_MOODLE, ['noclean' => false, 'trusted' => false, 'filter' => true, 'context' => $context, 'param' => true, 'newline' => true, 'allowid' => false, 'blanktarget' => false])
```

Comment:
I really don't think we should include funciton signatures in here. They change.

If we _do_, then we should format them in a readable way, and we should encourage the use of named parameters where possible.

```suggestion
function format_text(
    string $text,
    $format = FORMAT_MOODLE,
    [
        'noclean' => false,
        'trusted' => false,
        'filter' => true,
        'context' => $context,
        'param' => true,
        'newline' => true,
        'allowid' => false,
        'blanktarget' => false,
    ],
);

```

Reply:
We could also start to recommend that people consider using `\core\formatting` directly instead, e.g.

```php
\core\di::get(\core\formatting::class)->format_text(
    ?string $text,
    string $format = FORMAT_MOODLE,
    ?context $context = null,
    bool $trusted = false,
    ?bool $clean = null,
    bool $filter = true,
    bool $para = true,
    bool $newlines = true,
    bool $overflowdiv = false,
    bool $blanktarget = false,
    bool $allowid = false,
): string;
```